scripts/backout_data_collection.py

```diff
@@ -0,0 +1,198 @@
+        with open(bugzilla.BUGS_DB, "r") as f:
+            for line in tqdm(f, total=num_lines, desc="Processing bugs"):
+                bug = json.loads(line)
+                bug_dict[bug.get("id")] = bug["resolution"]
+
+        save_dict_to_file(commit_dict, cache_path_commit)
+        save_dict_to_file(bug_to_commit_dict, cache_path_bug_to_commit)
+        save_dict_to_file(bug_dict, cache_path_bug)
+
+    return commit_dict, bug_to_commit_dict, bug_dict
+
+
+def filter_commits(
+    directory_path: str,
+    destination_filepath: str,
+    count_limit: int,
+    bug_dict: dict,
+    commit_dict: dict,
+    bug_to_commit_dict: dict,
+) -> None:
+    filtered_list = []
+    counter = 0
+
+    pbar = tqdm(total=count_limit, desc="Filtering commits")
+
+    for commit in repository.get_commits(
+        include_no_bug=False, include_backouts=True, include_ignored=False
+    ):
+        # add commit if it was backed out and the bug is fixed
+        if (
+            bug_dict.get(str(commit["bug_id"]))
+            and commit["backedoutby"]
+            and bug_dict[str(commit["bug_id"])] == "FIXED"
+        ):
+            fixing_commit = find_next_commit(
+                commit["bug_id"], bug_to_commit_dict, commit["node"]
+            )
+
+            # if fixing commit could not be found or is another backing out commit, do not add it to dataset
+            if (
+                fixing_commit["node"] == commit["backedoutby"]
+                or len(fixing_commit["backsout"]) > 0
+            ):
+                counter += 1
+                continue
+
+            # add the hashes of the bug-inducing commit, the back out commit, and the fixing commit
+            # include metadata such as push date and description for further context
+            list_entry = {
+                "bug_id": commit["bug_id"],
+                "inducing_commit": {
+                    "node": commit["node"],
+                    "pushdate": commit["pushdate"],
+                    "desc": commit["desc"],
+                },
+                "backout_commit": {
+                    "node": commit["backedoutby"],
+                    "pushdate": commit_dict[commit["backedoutby"]]["pushdate"],
+                    "desc": commit_dict[commit["backedoutby"]]["desc"],
+                },
+                "fixing_commit": {
+                    "node": fixing_commit["node"],
+                    "pushdate": fixing_commit["pushdate"],
+                    "desc": fixing_commit["desc"],
+                },
+            }
+
+            filtered_list.append(list_entry)
+            counter += 1
+            pbar.update(1)
+
+            if counter >= count_limit:
+                break
+
+    json_data = json.dumps(filtered_list, indent=4)
+
+    if not os.path.exists(directory_path):
+        os.makedirs(directory_path)
+        print(f"Directory {directory_path} created")
+
+    with open(destination_filepath, "w") as file:
+        file.write(json_data)
```

Comment:
For the function called `filter_commits`, thus, I would avoid storing the data to the desc within the same function, which is an unexpected side effect.